value/atoms.go

```diff
@@ -9,12 +10,18 @@ import (
 // Nil represents the Value 'nil'.
 type Nil struct{}
 
+// SExpr returns a valid s-expression representing Nil.
+func (Nil) SExpr() (string, error) { return "nil", nil }
+
 // HashCode always returns 0.
 func (n Nil) HashCode() int64 { return 0 }
```

Comment:
Oh yea, we could go down that route. Actually, the `HashCode` function was something i was experimenting on, in Zulu based on Java spec. This was not meant to be part of parens. I guess we copied the value package from there. so it ended up here.

Reply:
>I guess we copied the value package from there. so it ended up here.

Yeah, I think that was me.  

Go's [`Hash`](https://golang.org/pkg/hash/#Hash) package works with `[]byte` as input, so I think we should define a `Hashable() []byte` method.

I'm inclined to make this part of the `value.Any` contract as well, since this is going to be required for things like hashmap types.  I'd rather not pay the price of type assertion on something so fundamental.

Reply:
Yea, i like the `Hashable() []byte` method. Not sure about adding it for all values though (since in our last discussion we decided not to do that since it's not possible in Go to get some hashable value for all value types)

Reply:
I am removing the HashCode from all values for now. Will add an issue for hashable types.

Reply:
>(since in our last discussion we decided not to do that since it's not possible in Go to get some hashable value for all value types)

Maybe I'm missing your point, but the advantage of a `Hashable()` method is that it gives you a way to produce a hashable representation of types that aren't hashable in native Go.  As an example, let's imagine we've defined a Parens type that satisfies `value.Any`, that is actually a `[]string` under the hood (this example is loosely based on Wetware's `Path` type).

```go
type Path []string

func (p Path) SExpr() (string, error) {
    var b strings.Builder
    b.WriteString("/")
    for _, s := range p {
        b.WriteString(s + "/")
    }
    return b.String(), nil
}
```

Now `[]string` is an unhashable type, but `Hashable()` lets us hash it anyway.  Here's a dead-simple implementation:

```go
func (p Path) Hashable() []byte {
    var b strings.Builder
    b.WriteString("/")
    for _, s := range p {
        b.WriteString(s + "/")
    }
    return b.Bytes()
}
```

🎉 Now you can hash a `[]string`!

Of course, if I've completely missed your point, please let me know 😅

Reply:
[]string works great. I'm thinking about the more native values like Func types for example. We should be able to expose go func values or some arbitrary struct values into parens env somehow so that they can be called/used from within lisp code. But if we expect every value type to have hashable, then we need to find a way that consistently returns same bytes for a specific value.. I don't think there's a good way to do that.

Reply:
> I'm thinking about the more native values like Func types for example.

This is effectively the defunctionalization/refunctionalization problem.  The trick is to substitute a symbol in lieu of the function object itself.  Here's a naïve solution for a native go `func()`:

```go
type GoFunc func()

func (f GoFunc) Hashable() []byte {
    return []byte(reflect.ValueOf(f).String())
}
```

The `reflect.Valueof(...).String()` call can be substituted with a lookup table as well.